src/main/kotlin/com/google/protobuf/gradle/ProtobufConfiguratorExts.kt

```diff
@@ -4,61 +4,170 @@ import org.gradle.api.NamedDomainObjectContainer
 import org.gradle.api.Project
 import org.gradle.api.plugins.ExtensionAware
 import org.gradle.api.tasks.SourceSet
-import org.gradle.kotlin.dsl.*
-
+import org.gradle.kotlin.dsl.NamedDomainObjectContainerScope
+import org.gradle.kotlin.dsl.closureOf
+import org.gradle.kotlin.dsl.get
 
+/**
+ * Applies the supplied action to the project's instance of [ProtobufConfigurator].
+ *
+ * @since 0.8.7
+ * @usage
+ * ```
+ * protobuf {
+ *     ...
+ *     generatedFilesBaseDir = files(...)
+ * }
+ * ```
+ *
+ * @receiver [Project] The project for which the plugin configuration will be applied
+ * @param action A configuration lambda to apply on a receiver of type [ProtobufConfigurator]
+ *
+ * @return [Unit]
+ */
 fun Project.protobuf(action: ProtobufConfigurator.()->Unit) {
     project.convention.getPlugin(ProtobufConvention::class.java).protobuf.apply(action)
 }
 
+/**
+ * Applies the supplied action to the [ProtobufSourceDirectorySet] extension on
+ * a receiver of type [SourceSet]
+ *
+ * @since 0.8.7
+ * @usage
+ * ```
+ * sourceSets {
+ *     create("sample") {
+ *         proto {
+ *             srcDir("src/sample/protobuf")
+ *         }
+ *     }
+ * }
+ * ```
+ *
+ * @receiver [SourceSet] The source set for which the [ProtobufSourceDirectorySet] extension
+ * will be configured
+ *
+ * @param action A configuration lambda to apply on a receiver of type [ProtobufSourceDirectorySet]
+ * @return [Unit]
+ */
 fun SourceSet.proto(action: ProtobufSourceDirectorySet.() -> Unit) {
     (this as? ExtensionAware)
         ?.extensions
         ?.getByType(ProtobufSourceDirectorySet::class.java)
         ?.apply(action)
 }
 
-fun ProtobufConfigurator.protoc(closure: ExecutableLocator.() -> Unit) {
-    protoc(closureOf(closure))
+/**
+ * Uses the supplied action to configure the [ExecutableLocator] for protoc.
+ *
+ * @since 0.8.7
+ * @usage
+ * ```
+ * protobuf {
+ *     protoc {
+ *         artifact = "com.google.protobuf:protoc:3.6.1"
+ *     }
+ * }
+ * ```
+ *
+ * @receiver [ProtobufConfigurator] The protobuf plugin configuration instance
+ * for the project.
+ *
+ * @param action A lambda with receiver of type [ExecutableLocator] used
+ * for configuring the locator for protoc
+ *
+ * @return [Unit]
+ */
+fun ProtobufConfigurator.protoc(action: ExecutableLocator.() -> Unit) {
+    protoc(closureOf(action))
 }
 
-fun ProtobufConfigurator.plugins(closure: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit) {
+fun ProtobufConfigurator.plugins(action: NamedDomainObjectContainerScope<ExecutableLocator>.() -> Unit) {
     plugins(closureOf<NamedDomainObjectContainer<ExecutableLocator>> {
-        this(closure)
+        NamedDomainObjectContainerScope.of(this).action()
```

Comment:
I changed the syntax usage in these methods to improve readability. At first I used the syntax from the code generated from gradle's plugin resolution stage for kotlin dsl. In hind sight, it was confusing to interpret for those not familiar with kotlin.